cmd/run/logger/logger.go

```diff
@@ -0,0 +1,91 @@
+package logger
+
+import (
+	"bufio"
+	"io"
+	"sync"
+)
+
+var Enabled bool
+
+var bufferSize = 4096
+
+type Logger struct {
+	wmu sync.Mutex
+	pr  *io.PipeReader
+	pw  *io.PipeWriter
+
+	rmu     sync.RWMutex
+	lines   []string
+	current int
+}
+
+func New() *Logger {
+	pr, pw := io.Pipe()
+	return &Logger{
+		lines:   make([]string, bufferSize),
+		current: 0,
+
+		pr: pr,
+		pw: pw,
+	}
+}
+
+func (l *Logger) Loop() {
+	s := bufio.NewScanner(l.pr)
+	defer l.pr.Close()
+	defer l.pw.Close()
+
+	for s.Scan() {
+		line := s.Text()
+		if len(line) > 1024 {
+			line = line[:1024]
+		}
+
+		l.addLine(line)
+	}
+}
+
+func (l *Logger) addLine(line string) {
+	l.rmu.Lock()
+	defer l.rmu.Unlock()
+
+	l.lines[l.current%bufferSize] = line
+	l.current++
+}
+
+func (l *Logger) GetLogs(start int) []string {
+	l.rmu.RLock()
+	defer l.rmu.RUnlock()
+
+	numLines := l.current - start
+	var result []string
+
+	if numLines <= bufferSize {
+		bstart := start % bufferSize
+		bend := l.current % bufferSize
+
+		if bstart <= bend {
+			result = make([]string, bend-bstart+1)
+			copy(result, l.lines[bstart:bend+1])
```

Comment:
use `return append([]string{}, l.lines[bstart:bend+1]...)`, you won't need a `result` variable after that